binderhub/repoproviders.py

```diff
@@ -64,6 +66,36 @@ def get_build_slug(self):
         return '{user}-{repo}'.format(user='Rick', repo='Morty')
 
 
+class GitRepoProvider(RepoProvider):
+    """Bare bones git repo provider. Users must provide a resolve_ref in the following form
+
+    http://hostname.com/path/repo_path#sha1
+    """
+
+    name = Unicode("git")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        parsed = urllib.parse.urlparse(self.spec)
+        if not parsed.fragment:
+            raise ValueError("spec used for raw git api must include the sha1 as a fragment")
+        self.resolved_ref = parsed.fragment
+        self.repo = urllib.parse.urlunparse(parsed._replace(fragment=''))
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        return self.resolved_ref
+
+    def get_repo_url(self):
+        return self.repo
+
+    def get_build_slug(self):
+        parsed = urllib.parse.urlparse(self.spec)
+        parts = [parsed.netloc]
+        parts.extend(parsed.path.split(posixpath.sep))
+        return '-'.join(parts)
```

Comment:
There's an ambiguity here, because replacing `/` with `-` gives `foo-bar/baz` the same slug as `foo/bar-baz`. Some form of escaping would need to be done to avoid collisions.